File: app/src/main/java/com/example/quicknotes/view/ManageTagsFragment.java
package com.example.quicknotes.view;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import android.content.Context;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.text.InputType;
import android.widget.EditText;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceScreen;

import com.example.quicknotes.model.NoteViewModel;
import com.example.quicknotes.model.Tag;
import com.example.quicknotes.model.TagColorManager;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

public class ManageTagsFragment extends PreferenceFragmentCompat {
    private NoteViewModel noteViewModel;
    private int[] colorResIds;
    private String[] colorNames;

    @Override
    public void onViewCreated(@NonNull android.view.View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        noteViewModel = new ViewModelProvider(requireActivity()).get(NoteViewModel.class);
        NavHostFragment.findNavController(this);
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        Context ctx = requireContext();
        noteViewModel = new ViewModelProvider(requireActivity()).get(NoteViewModel.class);

        List<TagColorManager.ColorOption> colorOptions = noteViewModel.getTagManager().getAvailableColors();
        Set<Tag> allTags = noteViewModel.getTagManager().getAllTags();

        colorResIds = new int[colorOptions.size()];
        colorNames = new String[colorOptions.size()];
        for (int i = 0; i < colorOptions.size(); i++) {
            TagColorManager.ColorOption option = colorOptions.get(i);
            colorResIds[i] = option.resId();
            colorNames[i] = option.name();
        }

        PreferenceScreen screen = getPreferenceManager().createPreferenceScreen(ctx);
        setPreferenceScreen(screen);

        if (allTags.isEmpty()) {
            Preference noTagsPref = new Preference(ctx);
            noTagsPref.setTitle("No tags available");
            noTagsPref.setSummary("Create tags by editing notes");
            noTagsPref.setSelectable(false);
            screen.addPreference(noTagsPref);
        } else {
            List<Tag> tagList = new ArrayList<>(allTags);
            tagList.sort((a, b) -> a.name().compareToIgnoreCase(b.name()));
            for (Tag tag : tagList) {
                screen.addPreference(createTagPreference(ctx, tag));
            }
        }

        Preference mergePref = new Preference(ctx);
        mergePref.setTitle("Merge Tags");
        mergePref.setSummary("Combine multiple tags into one");
        mergePref.setOnPreferenceClickListener(p -> {
            showMergeDialog();
            return true;
        });
        screen.addPreference(mergePref);
    }

    private Preference createTagPreference(Context ctx, Tag tag) {
        Preference pref = new Preference(ctx);
        String tagName = tag.name();
        pref.setKey("manage_tag_" + tagName);
        pref.setTitle(tagName);
        int currentColorRes = tag.colorResId();
        int currentColor = ContextCompat.getColor(ctx, currentColorRes);
        pref.setIcon(new ColorDrawable(currentColor));
        pref.setIconSpaceReserved(true);
        pref.setSummary("Tap to rename, delete, or change color");
        pref.setOnPreferenceClickListener(p -> {
            showTagOptionsDialog(tagName, pref);
            return true;
        });
        return pref;
    }

    private void showTagOptionsDialog(String tagName, Preference pref) {
        String[] options = new String[]{"Rename", "Delete", "Change Color"};
        new MaterialAlertDialogBuilder(requireContext())
                .setTitle("Manage '" + tagName + "'")
                .setItems(options, (d, which) -> {
                    switch (which) {
                        case 0 -> showRenameDialog(tagName);
                        case 1 -> showDeleteConfirm(tagName);
                        case 2 -> showColorPicker(tagName, pref);
                    }
                })
                .show();
    }

    private void showRenameDialog(String oldName) {
        EditText input = new EditText(requireContext());
        input.setInputType(InputType.TYPE_CLASS_TEXT);
        input.setText(oldName);
        new MaterialAlertDialogBuilder(requireContext())
                .setTitle("Rename Tag")
                .setView(input)
                .setNegativeButton(android.R.string.cancel, null)
                .setPositiveButton(android.R.string.ok, (d, w) -> {
                    String newName = input.getText() != null ? input.getText().toString().trim() : "";
                    if (!newName.isEmpty()) {
                        noteViewModel.getTagManager().renameTag(oldName, newName);
                        refreshScreen();
                    }
                })
                .show();
    }

    private void showDeleteConfirm(String tagName) {
        new MaterialAlertDialogBuilder(requireContext())
                .setTitle("Delete Tag")
                .setMessage("Remove tag '" + tagName + "' from all notes?")
                .setNegativeButton(android.R.string.cancel, null)
                .setPositiveButton(android.R.string.ok, (d, w) -> {
                    noteViewModel.getTagManager().deleteTag(tagName);
                    refreshScreen();
                })
                .show();
    }

    private void showColorPicker(String tagName, Preference pref) {
        new MaterialAlertDialogBuilder(requireContext())
                .setTitle("Select color for '" + tagName + "'")
                .setItems(colorNames, (d, which) -> {
                    int chosen = colorResIds[which];
                    noteViewModel.getTagManager().setTagColor(tagName, chosen);
                    int color = ContextCompat.getColor(requireContext(), chosen);
                    pref.setIcon(new ColorDrawable(color));
                })
                .show();
    }

    private void showMergeDialog() {
        Set<Tag> allTags = noteViewModel.getTagManager().getAllTags();
        if (allTags.isEmpty()) return;
        List<String> names = new ArrayList<>();
        for (Tag t : allTags) names.add(t.name());
        names.sort(String::compareToIgnoreCase);

        String[] items = names.toArray(new String[0]);
        boolean[] checked = new boolean[items.length];
        new MaterialAlertDialogBuilder(requireContext())
                .setTitle("Select tags to merge")
                .setMultiChoiceItems(items, checked, (d, which, isChecked) -> checked[which] = isChecked)
                .setNegativeButton(android.R.string.cancel, null)
                .setPositiveButton("Next", (d, w) -> {
                    List<String> selected = new ArrayList<>();
                    for (int i = 0; i < items.length; i++) if (checked[i]) selected.add(items[i]);
                    if (selected.isEmpty()) return;
                    showMergeTargetDialog(selected);
                })
                .show();
    }

    private void showMergeTargetDialog(List<String> sources) {
        EditText input = new EditText(requireContext());
        input.setInputType(InputType.TYPE_CLASS_TEXT);
        new MaterialAlertDialogBuilder(requireContext())
                .setTitle("Merge into tag")
                .setView(input)
                .setNegativeButton(android.R.string.cancel, null)
                .setPositiveButton("Merge", (d, w) -> {
                    String target = input.getText() != null ? input.getText().toString().trim() : "";
                    if (target.isEmpty()) return;
                    Set<String> unique = new LinkedHashSet<>(sources);
                    unique.removeIf(s -> s.equalsIgnoreCase(target));
                    if (unique.isEmpty()) return;
                    noteViewModel.getTagManager().mergeTags(new ArrayList<>(unique), target);
                    refreshScreen();
                })
                .show();
    }

    private void refreshScreen() {
        onCreatePreferences(null, null);
    }
}

File: app/src/main/java/com/example/quicknotes/view/ManageTagsFragment.kt
package com.example.quicknotes.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.quicknotes.databinding.ItemManageTagBinding
import com.example.quicknotes.databinding.SheetManageTagsBinding
import com.example.quicknotes.model.tag.Tag
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.dialog.MaterialAlertDialogBuilder

/**
 * Bottom sheet to manage tags (color, rename, delete) from the search screen.
 */
class ManageTagsFragment : BottomSheetDialogFragment() {
    private var binding: SheetManageTagsBinding? = null
    private var listener: NotesUI.Listener? = null
    private val adapter = TagListAdapter()

    fun setListener(l: NotesUI.Listener?) { listener = l }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = SheetManageTagsBinding.inflate(inflater, container, false)
        return binding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding?.recycler?.layoutManager = LinearLayoutManager(requireContext())
        binding?.recycler?.adapter = adapter
        binding?.closeButton?.setOnClickListener { dismiss() }
        loadTags()
    }

    private fun loadTags() {
        val tags = (listener?.onGetAllTags() ?: mutableSetOf()).toList().sortedBy { it.name }
        adapter.updateData(tags)
    }

    private inner class TagListAdapter : RecyclerView.Adapter<TagListAdapter.VH>() {
        private var items: List<Tag> = emptyList()

        fun updateData(newData: List<Tag>) {
            val diff = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
                override fun getOldListSize() = items.size
                override fun getNewListSize() = newData.size
                override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean =
                    items[oldItemPosition].name.equals(newData[newItemPosition].name, true)
                override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
                    val a = items[oldItemPosition]; val b = newData[newItemPosition]
                    return a.name == b.name && a.colorResId == b.colorResId
                }
            })
            items = newData
            diff.dispatchUpdatesTo(this)
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH =
            VH(ItemManageTagBinding.inflate(LayoutInflater.from(parent.context), parent, false))

        override fun getItemCount(): Int = items.size

        override fun onBindViewHolder(holder: VH, position: Int) = holder.bind(items[position])

        inner class VH(private val b: ItemManageTagBinding) : RecyclerView.ViewHolder(b.root) {
            fun bind(tag: Tag) {
                b.tagName.text = tag.name
                b.colorDot.background.setTint(ContextCompat.getColor(requireContext(), tag.colorResId))
                b.root.setOnClickListener { showActions(tag) }
            }
        }
    }

    private fun showActions(tag: Tag) {
        val actions = arrayOf("Change color", "Rename", "Delete")
        MaterialAlertDialogBuilder(requireContext())
            .setTitle(tag.name)
            .setItems(actions) { _, which ->
                when (which) {
                    0 -> showColorPicker(tag)
                    1 -> promptRename(tag)
                    2 -> confirmDelete(tag)
                }
            }
            .show()
    }

    private fun showColorPicker(tag: Tag) {
        val options = listener?.onGetAvailableColors()?.filterNotNull() ?: return
        val names = options.map { it.name }.toTypedArray()
        val res = options.map { it.resId }.toIntArray()
        MaterialAlertDialogBuilder(requireContext())
            .setTitle("Select color for '${tag.name}'")
            .setItems(names) { _, idx ->
                listener?.onSetTagColor(tag.name, res[idx])
                loadTags()
            }
            .show()
    }

    private fun promptRename(tag: Tag) {
        val input = android.widget.EditText(requireContext())
        input.setText(tag.name)
        MaterialAlertDialogBuilder(requireContext())
            .setTitle("Rename tag")
            .setView(input)
            .setNegativeButton(android.R.string.cancel, null)
            .setPositiveButton("Rename") { _, _ ->
                val newName = input.text?.toString()?.trim().orEmpty()
                if (newName.isNotEmpty() && !newName.equals(tag.name, true)) {
                    listener?.onRenameTag(tag.name, newName)
                    loadTags()
                }
            }
            .show()
    }

    private fun confirmDelete(tag: Tag) {
        MaterialAlertDialogBuilder(requireContext())
            .setTitle("Delete tag")
            .setMessage("Remove '${tag.name}' from all notes?")
            .setNegativeButton(android.R.string.cancel, null)
            .setPositiveButton("Delete") { _, _ ->
                listener?.onDeleteTag(tag.name)
                loadTags()
            }
            .show()
    }
}


